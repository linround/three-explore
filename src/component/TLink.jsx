import { useState } from 'react'
import PropTypes from 'prop-types'

const STATUS = {
  HOVERED: 'hovered',
  NORMAL: 'normal',
}

export default function TLink({ page, children, }) {
  const [status, setStatus] = useState(STATUS.NORMAL)

  const onMouseEnter = () => {
    setStatus(STATUS.HOVERED)
  }

  const onMouseLeave = () => {
    setStatus(STATUS.NORMAL)
  }

  return (
    <a
      className={status}
      href={page || '#'}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
    >
      {children}
    </a>
  )
}
TLink.propTypes  = {
  page: PropTypes.string,
  children: PropTypes.elementType,
}
