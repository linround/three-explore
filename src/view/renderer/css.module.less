.container {
  width: 100vw;
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
  position: relative;
}
.canvas {
  width: 600px;
  height: 600px;
  border-radius: 16px;
  border: 1px solid black;
}
