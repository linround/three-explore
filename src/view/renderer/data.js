export const sceneData = {
  // 观察点数据
  observationPoint: {
    position: {
      x: -800,
      y: -800,
      z: -800,
    },
    outPutNode: 'observationPointOutPutNode',
  },
  // 目标点坐标
  targetPoint: {
    position: {
      x: -200,
      y: -200,
      z: 600,
    },
    outPutNode: 'targetPointOutPutNode',
  },
  // 目标点-观察点 得到观察方向向量
  observationDirection: {
    position: {
      x: -1000,
      y: -1000,
      z: 1400,
    },
    inPutNode: 'observationDirectionInPutNode',
  },
}
