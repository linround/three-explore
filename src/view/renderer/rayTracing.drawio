<mxfile host="app.diagrams.net" modified="2023-10-19T06:31:25.106Z" agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36" version="22.0.5" etag="HOCyeEM6I9sPmE0JCBZ5" type="github">
  <diagram name="第 1 页" id="G4zZYvhnWSBs3zv3X2oq">
    <mxGraphModel>
      <root>
        <mxCell id="0" />
        <mxCell id="1" parent="0" />
      </root>
    </mxGraphModel>
  </diagram>
</mxfile>
