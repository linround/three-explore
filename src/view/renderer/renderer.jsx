import { Component, createRef } from 'react'
import cornellBoxFragShader from './cornellBoxFragShader.glsl?raw'
import commonVertexShader from '../../common/commonVertexShader.glsl?raw'
import * as THREE from 'three'
import bayer from '../../assets/bayer.png'
import { resizeRendererToDisplaySize } from '../../utils.js'
import css from './css.module.less'
import { Layer } from './layer/layer.jsx'

export class Renderer extends Component {
  constructor(props) {
    super(props)
    this.canvas = createRef()
    this.renderScene = this.renderScene.bind(this)
    this.renderCornellBox = this.renderCornellBox.bind(this)
  }
  componentDidMount() {
    this.renderCornellBox()
  }

  renderCornellBox() {
    this.renderScene(cornellBoxFragShader, commonVertexShader)

  }
  renderScene(fragmentShader, vertexShader) {
    const canvas = this.canvas.current
    const renderer = new THREE.WebGLRenderer({ canvas, })
    const camera = new THREE.OrthographicCamera(
      -1, 1, 1, -1, -1, 1
    )
    const scene = new THREE.Scene()
    const plane = new THREE.PlaneGeometry(2, 2)
    const loader = new THREE.TextureLoader()
    const texture = loader.load(bayer)
    texture.minFilter = THREE.NearestFilter
    texture.magFilter = THREE.NearestFilter
    texture.wrapS = THREE.RepeatWrapping
    texture.wrapT = THREE.RepeatWrapping
    const uniforms = {
      iFrame: 2,
      iTime: { value: 0, },
      iResolution: { value: new THREE.Vector3(), },
      iMouse: { value: new THREE.Vector2(), },
      iChannel0: { value: texture, },
    }
    const material = new THREE.ShaderMaterial({
      fragmentShader,
      vertexShader,
      uniforms,
    })

    scene.add(new THREE.Mesh(plane, material))
    let x = 0
    let y = 0

    const render = (t) => {
      const time = 0.001 * t
      resizeRendererToDisplaySize(renderer)
      uniforms.iResolution.value.set(
        canvas.width, canvas.height, 1
      )
      uniforms.iMouse.value.set(x, y)
      uniforms.iTime.value = time

      renderer.render(scene, camera)
      requestAnimationFrame(render)

    }
    requestAnimationFrame(render)
    canvas.addEventListener('pointermove', (e) => {
      x = e.offsetX
      y = canvas.height - e.offsetY
    })
  }
  render() {
    return (
      <div className={css.container}>
        <canvas ref={this.canvas}className={css.canvas} />
        <Layer />
      </div>
    )
  }
}
