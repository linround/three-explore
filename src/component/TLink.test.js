import renderer from 'react-test-renderer'
import TLink from './TLink'

it('hover 状态改变类名', () => {
  const component = renderer.create(<TLink page="https://three.ucalendar.cn/">three</TLink>)
  let tree = component.toJSON()
  expect(tree)
    .toMatchSnapshot()

  // 手动触发onMouseEnter
  renderer.act(() => {
    tree.props.onMouseEnter()
  })
  // 再次渲染
  tree = component.toJSON()
  expect(tree)
    .toMatchSnapshot()

  // 手动触发 onMouseLeave
  renderer.act(() => {
    tree.props.onMouseLeave()
  })
  // 再次渲染
  tree = component.toJSON()
  expect(tree)
    .toMatchSnapshot()
})
