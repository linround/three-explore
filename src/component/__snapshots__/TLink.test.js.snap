// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hover 状态改变类名 1`] = `
<a
  className="normal"
  href="https://three.ucalendar.cn/"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  three
</a>
`;

exports[`hover 状态改变类名 2`] = `
<a
  className="hovered"
  href="https://three.ucalendar.cn/"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  three
</a>
`;

exports[`hover 状态改变类名 3`] = `
<a
  className="normal"
  href="https://three.ucalendar.cn/"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  three
</a>
`;
